homeassistant/components/rainforest_raven/config_flow.py

```diff
@@ -0,0 +1,159 @@
+"""Config flow for Rainforest RAVEn devices."""
+from __future__ import annotations
+
+import asyncio
+from typing import Any
+from xml.etree.ElementTree import ParseError
+
+from aioraven.serial import RAVEnSerialDevice
+import async_timeout
+from serial.serialutil import SerialException
+import serial.tools.list_ports
+from serial.tools.list_ports_common import ListPortInfo
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.components import usb
+from homeassistant.const import CONF_DEVICE, CONF_MAC, CONF_NAME
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import DEFAULT_NAME, DOMAIN
+
+
+def _generate_unique_id(port: ListPortInfo) -> str:
+    """Generate unique id from usb attributes."""
+    return f"{port.vid:04X}:{port.pid:04X}_{port.serial_number}_{port.manufacturer}_{port.description}"
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Rainforest RAVEn devices."""
+
+    def __init__(self) -> None:
+        """Set up flow instance."""
+        self._dev_path: str | None = None
+        self._meter_macs: dict[bytes, str] = {}
+
+    async def _validate_device(self, dev_path: str) -> None:
+        self._abort_if_unique_id_configured(updates={CONF_DEVICE: dev_path})
+        async with (
+            async_timeout.timeout(5),
+            RAVEnSerialDevice(dev_path) as raven_device,
+        ):
+            await asyncio.sleep(0.05)
+            for _try in range(3, -1, -1):
+                try:
+                    # Try a few times to communicate with the device,
+                    # allowing any data already in the buffer to flush.
+                    meters = await raven_device.get_meter_list()
+                except ParseError:
+                    if not _try:
+                        raise
+                else:
+                    break
+            for meter in meters.meter_mac_ids:
+                meter_info = await raven_device.get_meter_info(meter=meter)
+                self._meter_macs[meter] = (
+                    "unknown"
+                    if meter_info.meter_type is None
+                    else str(meter_info.meter_type)
+                )
```

Comment:
This code should go into the third party library

Reply:
I moved the synchronization logic into the library. I'm not sure I agree that the rest of this code should go there. It's opening the device, getting the list of programmed meters, and getting the type of each of those meters. These are all discrete operations and I don't see much value in lumping them together in the third-party library and breaking the one-to-one relationship between library queries and device messages.